src/target/riscv/riscv.c

```diff
@@ -5073,59 +5072,65 @@ int riscv_enumerate_triggers(struct target *target)
-			if (result != ERROR_OK)
-				return result;
+			trigger_tinfo[t] = tinfo;
+		}
 
-			int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+		riscv_reg_t tdata1;
+		if (riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1) != ERROR_OK)
+			return ERROR_FAIL;
+
+		const int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+		if (!may_support_multiple_trigger_types) {
 			if (type == 0)
 				break;
-			switch (type) {
-				case CSR_TDATA1_TYPE_LEGACY:
-					/* On these older cores we don't support software using
-						* triggers. */
-					riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-					break;
-				case CSR_TDATA1_TYPE_MCONTROL:
-					if (tdata1 & CSR_MCONTROL_DMODE(riscv_xlen(target)))
-						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-					break;
-				case CSR_TDATA1_TYPE_MCONTROL6:
-					if (tdata1 & CSR_MCONTROL6_DMODE(riscv_xlen(target)))
-						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-					break;
-				case CSR_TDATA1_TYPE_ICOUNT:
-					if (tdata1 & CSR_ICOUNT_DMODE(riscv_xlen(target)))
-						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-					break;
-				case CSR_TDATA1_TYPE_ITRIGGER:
-					if (tdata1 & CSR_ITRIGGER_DMODE(riscv_xlen(target)))
-						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-					break;
-				case CSR_TDATA1_TYPE_ETRIGGER:
-					if (tdata1 & CSR_ETRIGGER_DMODE(riscv_xlen(target)))
-						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-					break;
-			}
-			r->trigger_tinfo[t] = 1 << type;
+			trigger_tinfo[t] = 1 << type;
 		}
-		LOG_TARGET_DEBUG(target, "Trigger %u: supported types (mask) = 0x%08x", t, r->trigger_tinfo[t]);
-	}
+		/* Disable leftover triggers */
+		bool is_leftover;
+		switch (type) {
+			case CSR_TDATA1_TYPE_LEGACY:
+				/* On these older cores we don't support software using
+				 * triggers. */
+				is_leftover = true;
+				break;
+			case CSR_TDATA1_TYPE_MCONTROL:
+				is_leftover = tdata1 & CSR_MCONTROL_DMODE(riscv_xlen(target));
+				break;
+			case CSR_TDATA1_TYPE_MCONTROL6:
+				is_leftover = tdata1 & CSR_MCONTROL6_DMODE(riscv_xlen(target));
+				break;
+			case CSR_TDATA1_TYPE_ICOUNT:
+				is_leftover = tdata1 & CSR_ICOUNT_DMODE(riscv_xlen(target));
+				break;
+			case CSR_TDATA1_TYPE_ITRIGGER:
+				is_leftover = tdata1 & CSR_ITRIGGER_DMODE(riscv_xlen(target));
+				break;
+			case CSR_TDATA1_TYPE_ETRIGGER:
+				is_leftover = tdata1 & CSR_ETRIGGER_DMODE(riscv_xlen(target));
+				break;
+		}
+		if (is_leftover && riscv_set_register(target, GDB_REGNO_TDATA1, 0) != ERROR_OK)
+			return ERROR_FAIL;
 
-	riscv_set_register(target, GDB_REGNO_TSELECT, tselect);
+		LOG_TARGET_DEBUG(target, "Trigger %u: supported types (mask) = 0x%08x",
+				t, trigger_tinfo[t]);
+	}
 
-	LOG_TARGET_INFO(target, "Found %d triggers.", r->trigger_count);
+	if (riscv_set_register(target, GDB_REGNO_TSELECT, tselect) != ERROR_OK)
+		return ERROR_FAIL;
 
+	r->triggers_enumerated = true;
+	r->trigger_count = t + 1;
```

Comment:
This is incorrect. Why do you do this increment?

Reply:
Addressed.